docs/spec/lightclient/detection.md

```diff
@@ -0,0 +1,487 @@
+discussion about forks) in case there is one.
+
+2) to support fork accountability: In the case when more than 1/3 of
+the voting power is held by faulty validators, faulty nodes may
+generate two conflicting headers for the same height. The goal of the
+detector is to learn about the conflicting headers by probing
+different full nodes. Once a detector has two conflicting headers,
+these headers are evidence of misbehavior. A natural extension is to
+use the detector within a monitor process (on a full node) that calls
+the detector on a sample (or all) headers (in parallel). (If the
+sample is chosen at random, this adds a level of probabilistic
+reasoning.) If conflicting headers are found, they are evidence that
+can be used for punishing processes.
+
+In this document we will focus onn strengthening the light client, and
+leave other uses of the detection mechanism (e.g., when run on a full
+node) to the future.
+
+
+## Context of this document
+
+The light client verification specification [[verification]] is
+designed for the Tendermint failure model (1/3 assumption)
+[TMBC-FM-2THIRDS]. It is safe under this assumption, and live
+if it can reliably (that is, no message loss, no duplication, and
+eventually delivered) and timely communicate with a correct full node. If
+this assumption is violated, the light client can be fooled to trust a
+header that was not generated by Tendermint consensus.
+
+This specification, the fork detector, is a "second line of defense",
+in case the 1/3 assumption is violated. Its goal is to collect
+evidence. However, it is impractical to probe all full nodes. At this
+time we consider a simple scheme of maintaining an address book of
+known full nodes from which a small subset (e.g., 4) are chosen
+initially to communicate with. More involved book keeping with
+probabilistic guarantees can be considered at later stages of the
+project.
+
+The light client maintains a simple address book containing addresses
+of full nodes that it can pick as primary and secondaries.  To obtain
+a new header, the light client first does [verification](verification)
+with the primary, and then cross-checks the header with the
+secondaries using this specification.
+
+
+### Tendermint Consensus and Forks
+
+#### **[TMBC-GENESIS]**
+Let *Genesis* be the agreed-upon initial block (file).
+
+#### **[TMBC-FUNC]**
+> **TODO** be more precise. +2/3 of b.NextV = c.Val signed c. For now
+> the following will do:
+
+Let b and c be two light blocks
+with *b.Header.Height + 1 = c.Header.Height*. We define **signs(b,c)**
+    iff `ValidAndVerified(b, c)` 
+
+> **TODO** be more precise. +1/3 of b.NextV signed c. For now
+> the following will do:
+
+Let *b* and *c* be two light blocks. We define **supports(b,c,t)** 
+    iff `ValidAndVerified(b, c)` at time *t*
+
+
+> The following formalizes that *b* was properly generated by
+> Tendermint; *b* can be traced back to genesis
+
+#### **[TMBC-SEQ-ROOTED]**
```

Comment:
Based on the definitions below it looks to me like [TMBC-SEQ-ROOTED] and [TMBC-SKIP-ROOTED] are the other way around.

Reply:
I am not sure to understand your comment. Is it about the labels?

Reply:
Sorry, the explanation is not that good. What I meant is that in the **TMBC-SEQ-ROOTED** **sequ-rooted** such that there exists a block of height *i* between 1 and *h* which supports `supports(a(i), a(i+1), t), for 1 <= i < h`. This definition in my opinion is closer to skipping instead of sequential as the naming would suggest.

Reply:
Ah! I read "for 1 <= i < h" as "for all i, 1 <= i < h". That's how Knuth writes it ;-) But if it is confusing, I will make this more clear.

Reply:
Oh, ok than it makes sense :D